### ISSUE TITLE
ㅍㅌㅇㅅ

### ISSUE THREAD

**Issue by nhk9680** [nhk9680/storage#13]
> 데이터셋은 `train`과 `test1`로 나누어져 있으며, jpg 포맷의 이미지입니다.
이를 클러스터링 방법론 중 하나인 SVM을 이용하여 분류하는 방법입니다.

---

## 학습 데이터 전처리

`train` 폴더에는 `cat.0.jpg`, `cat.0.jpg`, ... 와 같이 `정답 라벨(cat/dog).index.jpg`로 구성되어 있습니다.
이를 이미지 X와 라벨 데이터 Y로 나누어 저장합니다.

```
for i, img in enumerate(paths.list_images(dataset_train)):
```

`imutils` 라이브러리의 `paths` 클래스 중 `list_images` 메소드를 이용하여 `dataset_train`을 로드합니다.

학습량을 조절하기 위해 변수 `i`를 추가하여 일정량(1000)이 되면 중단되도록 설정하였습니다.

이상적으로는 test 데이터보다 수가 많아야 하지만, colab notebook의 runtime 문제 등으로 인하여 일부만 포함시켰습니다.

만약 본인의 환경이 여유가 된다면 모든 train 데이터를 포함시키는 것을 권장드립니다.

```
# dog: 1
  Y.append(1 if 'dog' in img else 0)  
  img = cv2.imread(img)
#  print(img.shape)
  img = cv2.resize(img, (32, 32), interpolation=cv2.INTER_AREA)
#  print(img.shape)
  img = img.flatten()
#  print(img.shape)
  X.append(img)
```

`cat`은 0, `dog`는 1로 매핑하였으므로 이미지 파일명을 확인하여 라벨을 추가합니다.

OpenCV를 이용하여 이미지 파일을 불러온 후, `resize()` 메소드를 이용하여 32x32 크기로 조정합니다.

마지막으로 용이한 학습을 위해 2차원 이미지 배열을 1차원으로 변환하여 이미지 배열 X에 추가합니다.

```
X = np.array(X)
Y = np.array(Y)
```

위에서 입력받은 `X`, `Y`는 type이 `list` 형태이므로, `array` 형태로 형변환 해줍니다.

이렇게 전처리된 데이터를 학습할 때, 오버피팅을 방지하고 어느 정도의 성능이 나오는지 확인하기 위해 train set과 validation set으로 분할합니다.

```
X_train, X_val, Y_train, Y_val = train_test_split(
    X, Y, test_size=0.25, random_state=42)
```

`scikit learn`에서 제공해주는 `train_test_split()` 메소드를 이용하여 간편하게 분할해줍니다. 이 때, test size는 25%를 권장합니다. 보통 8:2, 7:3 비율도 많이 이용합니다.

---

## 학습 기반 분류기 설계

앞서 말씀드린대로 SVM(Support Vector Machine)을 적용하기 위해, `scikit learn`에서 제공해주는 `SVC` 클래스를 이용합니다.

```
svc = SVC(kernel='poly')
svc.fit(X_train, Y_train)
```

적용 가능한 SVM의 `kernel`은 `linear`, `poly`, `rbf` 등이 있습니다. 기본값은 `rbf`입니다.

`svc.fit()`으로 모델을 학습시킵니다.

```
predict = svc.predict(X_val)
print(classification_report(Y_val, predict))
```

학습된 모델을 이용하여 validation set을 적용해 보고, `classification_report`를 이용하여 precision-recall 값과 mAP(mean Average Precision) 값을 구해봅니다.

```
              precision    recall  f1-score   support

           0       0.58      0.66      0.62       123
           1       0.62      0.54      0.58       128

    accuracy                           0.60       251
   macro avg       0.60      0.60      0.60       251
weighted avg       0.60      0.60      0.60       251
```
결과는 위와 같습니다.

---

## 테스트 데이터 전처리

앞서 진행했던 것과 동일하게, 이번에는 테스트 데이터를 전처리합니다.

다만 차이점은, 테스트 데이터는 정답이 없는 `index.jpg`의 형태입니다.
따라서 Y 배열은 필요없고 X 배열만 선언해줍니다.
또한, 테스트 데이터는 validation이 필요 없으므로 분할하지 않고, 모두 포함합니다.

---

## 학습한 모델로 테스트하기

```
result = svc.predict(X_test)
```
앞서 학습해둔 `svc`에 테스트 셋을 적용하여 그 결과를 추출합니다.

---

## csv로 저장하기

python에서 데이터 셋을 조금 더 편하게 다루게 해주는 `pandas` 라이브러리를 이용합니다. pandas에서는 `DataFrame`이라는 형태로 데이터를 처리합니다. 이는 table과 유사한 형태라고 생각하시면 됩니다.

```
df = pd.DataFrame(result, columns=['label'])
```
DataFrame을 생성하고, 그 안에 추출한 결과값을 대입합니다. 열의 이름은 `label`로 설정합니다.

```
df.index += 1
df.index.name = 'id'
```

제출 예시인 Sample csv파일의 규격을 맞추기 위해
- index 값의 시작을 1로 맞추고,
- index 열의 이름을 'id'로 설정해줍니다.

```
df.to_csv('result_namhun_kim-2320.csv',index=True, header=True)
```
마지막으로 `to_csv()` 메소드를 이용하여 DataFrame을 csv 포맷으로 저장합니다.

**Comment by nhk9680** [comment]
> 오디오의 비트, 템포 등 여러 음악적 요소들을 이용하여 모델을 학습하고, 장르에 맞게 분류합니다. 본 코드에서는 SVM(Support Vector Machine) 이라는 클러스터링 방법론 중 하나를 사용했습니다.

---

## 학습 데이터 전처리

python에서 데이터 셋을 조금 더 편하게 다루게 해주는 `pandas` 라이브러리를 이용합니다. pandas에서는 `DataFrame`이라는 형태로 데이터를 처리합니다. 이는 table과 유사한 형태라고 생각하시면 됩니다.

```
df_data = pd.read_csv('../input/2019-pr-midterm-musicclassification/data_train.csv')
df_data.drop(['filename'], axis=1, inplace=True)
```

`read_csv()`를 이용하여 csv로 전처리된 오디오 파일을 로드합니다.
그다음 학습에 불필요한 `filename` 열을 제거하기 위해 `drop()`을 이용합니다.

전처리된 데이터를 학습할 때, 오버피팅을 방지하고 어느 정도의 성능이 나오는지 확인하기 위해 train set과 validation set으로 분할합니다.

```
split = StratifiedShuffleSplit(test_size=0.25,  random_state=42)
```

로드한 csv 데이터셋은 장르에 따라 정렬되어 있으므로, 이를 그대로 train과 val로 나눈다면 몇몇 장르가 아예 포함되지 않을 수 있습니다. 따라서 무작위로 섞어서 나눠줄 필요가 있는데, 이를 `StratifiedShuffleSplit` 클래스를 이용하여 분할합니다.

```
for train_idx, val_idx in split.split(df_data, df_data["label"]):
    train = df_data.iloc[train_idx]
    val = df_data.iloc[val_idx]    
```

잘 섞인 데이터셋을, `split.split()`을 이용하여 학습할 오디오 소스와 정답 데이터인 라벨로 분리하고, `train`과 `val`로 나누어 줍니다. 첫 번째 파라미터가 X, 두 번째 파라미터가 Y입니다.

이 떄 특정 데이터만 골라내기 위해 인덱싱(indexing)을 해야 하는데, pandas에서는 인덱싱 메소드를 지원해줍니다.
`iloc()` 메소드는 순서를 나타내는 정수 기반인 2차원 인덱싱입니다. `loc()`와는 다르게, 라벨이 아닌 순서를 나타내는 정수, 즉 index만 받는다는 차이점이 있습니다.

```
X_train = train.drop('label', axis=1)
Y_train = train['label']

X_val = val.drop('label', axis=1)
Y_val = val['label']
```
앞서 사용했던 `drop()` 메소드를 이용하여, X에는 라벨을 제외한 데이터만, Y에는 라벨만 넣어줍니다.

```
X_train = scale(X_train)
X_val = scale(X_val)
```

전처리된 csv파일의 오디오 소스는 값의 범위가 파라미터별로 전부 다르고, 분포 또한 다양합니다. 따라서 outlier를 줄이고 안정적인 학습을 위해 `scale` 라이브러리를 이용하여 값을 정규화(normalization)합니다. 계산식은 매우 간단하며, 결과를 확인해보면 범위가 -1~1 사이로 압축된 것을 확인할 수 있습니다.

---

## 학습 기반 분류기 설계

앞서 말씀드린대로 SVM(Support Vector Machine)을 적용하기 위해, scikit learn에서 제공해주는 `SVC` 클래스를 이용합니다.

```
svc = SVC(kernel = 'poly')
svc.fit(X_train, Y_train)
```

적용 가능한 SVM의 `kernel`은 `linear`, `poly`, `rbf` 등이 있습니다. 기본값은 `rbf`입니다.

`svc.fit()`으로 모델을 학습시킵니다.

```
predict = svc.predict(X_val)
print(classification_report(Y_val, predict))
```

학습된 모델을 이용하여 validation set을 적용해 보고, `classification_report`를 이용하여 precision-recall 값과 mAP(mean Average Precision) 값을 구해봅니다.

```
              precision    recall  f1-score   support

       blues       0.69      0.48      0.56        23
   classical       1.00      0.79      0.88        24
     country       0.38      0.67      0.48        24
       disco       0.42      0.33      0.37        24
      hiphop       0.74      0.58      0.65        24
        jazz       0.74      0.71      0.72        24
       metal       0.79      0.79      0.79        24
         pop       0.86      0.75      0.80        24
      reggae       0.34      0.48      0.40        23
        rock       0.35      0.33      0.34        24

    accuracy                           0.59       238
   macro avg       0.63      0.59      0.60       238
weighted avg       0.63      0.59      0.60       238
```
결과는 위와 같습니다.

---

## GridSearch를 이용한 성능 향상

본 데이터셋은 학습 데이터의 개수가 950개로 매우 적은 편에 속합니다. 게다가 25%를 validation set으로 떼어내게 된다면 실질적으로 학습에 이용되는 데이터 수는 더 적집니다.
따라서 교차 검증(cross validation) 방법을 이용하여 validation 데이터까지 모두 학습에 활용합니다.
또한, 성능 향상을 위해 SVM의 `C`나 `gamma`같은 파라미터 값들을 조정해 가며 최적의 모델을 찾을 수도 있습니다.
이 두 가지 방법을 동시에 진행하는 것이 `GridSearch` 입니다.

```
param_grid = [{'kernel':["poly"], \
               'degree': [1,2,3,4,5],\
               'gamma': [0.01, 0.1, 0.5, 1], \
               'coef0': [0, 0.1, 0.5, 1], \
               'C': [0.001, 0.01, 0.05, 0.1, 0.5, 1, 5, 10]}]
```
SVM 모델에 적용할 파라미터 값들을 tuple 형태로 지정하여 줍니다.

```
grid_search = GridSearchCV(svc, param_grid, cv=10, n_jobs=-1)
grid_search.fit(X_val, Y_val)
```

`GridSearchCV` 클래스를 이용하여 최적화된 값을 찾습니다.

```
grid_search.best_params_
```

결과는 아래와 같이 출력됩니다.

```
{'C': 5, 'coef0': 1, 'degree': 4, 'gamma': 0.01, 'kernel': 'poly'}
```

이를 모델에 반영하여 다시 학습 후 성능을 확인합니다.

```
svc = SVC(C=5, kernel='poly', degree=4)
svc.fit(X_train, Y_train)
predict = svc.predict(X_val)
print(classification_report(Y_val, predict))
```
결과는 아래와 같습니다.
```
              precision    recall  f1-score   support

       blues       0.55      0.48      0.51        23
   classical       0.82      0.75      0.78        24
     country       0.42      0.58      0.49        24
       disco       0.40      0.33      0.36        24
      hiphop       0.50      0.42      0.45        24
        jazz       0.72      0.54      0.62        24
       metal       0.74      0.71      0.72        24
         pop       0.79      0.62      0.70        24
      reggae       0.26      0.48      0.33        23
        rock       0.25      0.21      0.23        24

    accuracy                           0.51       238
   macro avg       0.54      0.51      0.52       238
weighted avg       0.55      0.51      0.52       238
```

아쉽게도, 항상 최상의 성능값이 도출되는 것은 아닙니다.

---

## 테스트 데이터 전처리

```
df_data = pd.read_csv('../input/2019-pr-midterm-musicclassification/data_test.csv')
df_data.drop(['filename'], axis=1, inplace=True)

X_test = df_data.drop(['label'], axis=1)

X_test = scale(X_test)
```

앞서 했던 학습 데이터 전처리 방법과 동일합니다. 차이점은 정답인 라벨 데이터 Y가 없다는 것 뿐입니다.

---

## 학습한 모델로 테스트하기

```
result = svc.predict(X_test)
```

앞서 학습해둔 `svc`에 테스트 셋을 적용하여 그 결과를 추출합니다.

---

## csv로 저장하기

```
df = pd.DataFrame(result, columns=['label'])


df = df.replace('blues', 0)
df = df.replace('classical', 1)
df = df.replace('country', 2)
df = df.replace('disco', 3)
df = df.replace('hiphop', 4)
df = df.replace('jazz', 5)
df = df.replace('metal', 6)
df = df.replace('pop', 7)
df = df.replace('reggae', 8)
df = df.replace('rock', 9)

df.index += 1
df.index.name = 'id'

df.to_csv('result_namhun_kim-2316.csv',index=True, header=True)
```

매핑해둔 값을 제출 규격에 맞게 복원하고, pandas의 DataFrame을 이용하여 csv로 저장합니다.
